pynbody/snapshot/gadgethdf.py

```diff
@@ -272,42 +477,16 @@ def _all_hdf_groups_in_family(self, fam):
             yield from self._hdf_files.iter_particle_groups_with_name(hdf_family_name)
 
 
-    def __init_file_map(self):
-        family_slice_start = 0
-
-        all_families_sorted = self._families_ordered()
-
-        self._gadget_ptype_slice = {} # will map from gadget particle type to location in pynbody logical file map
-
-        for fam in all_families_sorted:
-            family_length = 0
-
-
-            # A simpler and more readable version of the code below would be:
-            #
-            # for hdf_group in self._all_hdf_groups_in_family(fam):
-            #     family_length += hdf_group[self._size_from_hdf5_key].size
-            #
-            # However, occasionally we need to know where in the pynbody file map the gadget particle types lie.
-            # (Specifically this is used when loading subfind data.) So we need to expand that out a bit and also
-            # keep track of the slice for each gadget particle type.
-
-            ptype_slice_start = family_slice_start
-
-            for particle_type in self._family_to_group_map[fam]:
-
-                ptype_slice_len = 0
-                for hdf_group in self._hdf_files.iter_particle_groups_with_name(particle_type):
-                    ptype_slice_len += hdf_group[self._size_from_hdf5_key].size
-                self._gadget_ptype_slice[particle_type] = slice(ptype_slice_start, ptype_slice_start + ptype_slice_len)
-                family_length += ptype_slice_len
-                ptype_slice_start += ptype_slice_len
-
-            self._family_slice[fam] = slice(family_slice_start, family_slice_start + family_length)
-            family_slice_start += family_length
+    def __init_file_map(self, take):
+        self._array_loader = HDFArrayLoader(self._hdf_files, self._families_ordered(), self._family_to_group_map, take)
+        self._gadget_ptype_slice = self._array_loader._file_ptype_slice
+        self._family_slice = self._array_loader._family_slice_to_load
+        self._num_particles = self._array_loader._num_particles_to_load
 
-
-        self._num_particles = family_slice_start
+        # PartType3 is Tracer in Arepo, so we'd better to adjust the group mapping, removing empty groups
```

Comment:
I don't understand this comment. the below seems to check for empty groups and remove them. What does this have to do with tracers specifically?

This manipulation probably also belongs in `__init_family_map` rather than here